internal/interpreter/highlight.rs

```diff
@@ -3,251 +3,68 @@
-    // We have a sibling, so find the "deepest first_child"
-    if let Some(s) = item.next_sibling() {
-        next_item_down(&s)
-    } else if !item.is_root_item_of(root_item_tree) {
-        item.parent_item().unwrap() // We are not at this component's root!
-    } else {
-        // We are at the root of the tree: Start over, going for the
-        // deepest first child again!
-        next_item_down(item)
-    }
-}
-
-fn next_item_down(item: &ItemRc) -> ItemRc {
-    if let Some(child) = item.first_child() {
-        next_item_down(&child)
-    } else {
-        item.clone()
-    }
-}
-
-fn element_providing_item(component: &ErasedItemTreeBox, index: u32) -> Option<ElementRc> {
-    generativity::make_guard!(guard);
-    let c = component.unerase(guard);
-
-    c.description().original_elements.get(index as usize).cloned()
-}
-
-fn find_element_range(element: &ElementRc) -> Option<(Option<SourceFile>, TextRange)> {
-    let e = &element.borrow();
-    e.node.as_ref().and_then(|n| {
-        n.parent()
-            .filter(|p| p.kind() == i_slint_compiler::parser::SyntaxKind::SubElement)
-            .map_or_else(
-                || Some((n.source_file().cloned(), n.text_range())),
-                |p| Some((p.source_file().cloned(), p.text_range())),
-            )
-    })
-}
-
-fn map_range_to_line(
-    source_file: Option<SourceFile>,
-    range: TextRange,
-) -> (String, u32, u32, u32, u32) {
-    source_file.map_or_else(
-        || (String::new(), 0, 0, 0, 0),
-        |sf| {
-            let file_name = sf.path().to_string_lossy().to_string();
-            let (start_line, start_column) = sf.line_column(range.start().into());
-            let (end_line, end_column) = sf.line_column(range.end().into());
-            (file_name, start_line as u32, start_column as u32, end_line as u32, end_column as u32)
-        },
-    )
-}
-
-struct DesignModeState {
-    pub current_item: Option<ItemWeak>,
-}
-
-/// Use the last item we visited (if that covers the click area) or the root element
-fn find_start_item(
-    state: &RefCell<DesignModeState>,
+fn collect_highlight_data(
     component: &DynamicComponentVRc,
-    position: &LogicalPoint,
-) -> ItemRc {
-    let component_rc = VRc::into_dyn(component.clone());
-    state
-        .try_borrow()
-        .ok()
-        .and_then(|s| s.current_item.clone())
-        .and_then(|i| i.upgrade())
-        .filter(|i| item_contains(i, position, &component_rc))
-        .unwrap_or_else(|| ItemRc::new(component_rc, 0))
-}
-
-fn item_contains(item: &ItemRc, position: &LogicalPoint, item_tree: &ItemTreeRc) -> bool {
-    let offset = item.map_to_item_tree(LogicalPoint::default(), item_tree);
-    let geometry = item.geometry().translate(offset.to_vector());
-
-    geometry.contains(*position)
-}
-
-pub fn on_element_selected(
-    component_instance: &DynamicComponentVRc,
-    callback: Box<dyn Fn(&str, u32, u32, u32, u32)>,
-) {
+    elements: &Vec<std::rc::Weak<RefCell<Element>>>,
```

Comment:
clippy would probably tell you to take a ref to a slice instead of a ref to a vector.